testing/test_multi.py

```diff
@@ -246,8 +247,9 @@ def kill():
     assert execmodel.active_count() == active
 
 
+@pytest.mark.xfail(reason="active_count() has been broken for some time")
 def test_safe_terminate2(execmodel):
-    if execmodel.backend != "threading":
+    if execmodel.backend not in ("thread", "main_thread_only"):
         pytest.xfail(
             "execution model %r does not support task count" % execmodel.backend
         )
```

Comment:
Both `test_safe_terminate` and `test_safe_terminate2` have `active_count()` usage which has been broken for some time, and the `execmodel.backend != "threading"` comparison made it an xfail even for the "thread" backend because it should have been `execmodel.backend != "thread"`.  There's also a [python 3.12 atexit  issue that affects test_terminate_implicit_does_trykill as well](https://github.com/pytest-dev/execnet/pull/243/files#r1493993031). I have fixes but I'd like to get feedback before I include them in this MR since they're kind of off-topic.

Reply:
These are the local changes that I used to get the tests working and remove the xfail, and there are 4 parts:
* Use `__del__` instead of atexit in Group to avoid `RuntimeError: can't create new thread at interpreter shutdown` under python3.12.
* Return workerpool from safe_terminate for use in `active_count()` assertions.
* Simply use `assert workerpool.active_count() == 0` because `threading.active_count()` is not actually relevant for the purposes of this test.
* Increase safe_terminate timeout from 1 to 3 in test_safe_terminate, so that it uses the same timeout as test_safe_terminate2. Otherwise, I've observed that a timeout of 1 is not long enough `workerpool.active_count()` returns 10 instead of 0.
```diff
--- a/src/execnet/multi.py
+++ b/src/execnet/multi.py
@@ -6 +5,0 @@ Managing Gateway Groups and interactions with multiple channels.
-import atexit
@@ -42 +41,3 @@ class Group:
-        atexit.register(self._cleanup_atexit)
+
+    def __del__(self):
+        self._cleanup_atexit()
@@ -311,0 +313 @@ def safe_terminate(execmodel, timeout, list_of_paired_functions):
+    return workerpool
diff --git a/testing/test_multi.py b/testing/test_multi.py
index 3996a96..146fe58 100644
--- a/testing/test_multi.py
+++ b/testing/test_multi.py
@@ -103 +103 @@ class TestGroup:
-        assert atexitlist == [group._cleanup_atexit]
+        assert atexitlist == []
@@ -226 +225,0 @@ class TestGroup:
-@pytest.mark.xfail(reason="active_count() has been broken for some time")
@@ -232 +230,0 @@ def test_safe_terminate(execmodel):
-    import threading
@@ -234 +231,0 @@ def test_safe_terminate(execmodel):
-    active = threading.active_count()
@@ -243 +240 @@ def test_safe_terminate(execmodel):
-    safe_terminate(execmodel, 1, [(term, kill)] * 10)
+    workerpool = safe_terminate(execmodel, 3, [(term, kill)] * 10)
@@ -247 +244 @@ def test_safe_terminate(execmodel):
-    assert execmodel.active_count() == active
+    assert workerpool.active_count() == 0
@@ -250 +246,0 @@ def test_safe_terminate(execmodel):
-@pytest.mark.xfail(reason="active_count() has been broken for some time")
@@ -256 +251,0 @@ def test_safe_terminate2(execmodel):
-    import threading
@@ -258 +252,0 @@ def test_safe_terminate2(execmodel):
-    active = threading.active_count()
@@ -267 +261 @@ def test_safe_terminate2(execmodel):
-    safe_terminate(execmodel, 3, [(term, kill)] * 10)
+    workerpool = safe_terminate(execmodel, 3, [(term, kill)] * 10)
@@ -271 +265 @@ def test_safe_terminate2(execmodel):
-    assert threading.active_count() == active
+    assert workerpool.active_count() == 0
```
This is all kind of off-topic for the MR, so I'd like to get feedback before I include it.